### PR TITLE
Re-write sonic-db-cli with c++ for sonic startup performance issue

### PR DESCRIPTION
**Pull request by liuh-80**
#### Why I did it
    Fix sonic-db-cli high CPU usage on SONiC startup issue: https://github.com/Azure/sonic-buildimage/issues/10218
    ETA of this issue will be 2022/05/31

#### How I did it
    Re-write sonic-cli with c++.

#### How to verify it
    Add c++ unit test to cover all code.
    Pass all E2E test scenario.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog
    Re-write sonic-cli with c++ for sonic startup performance issue

### CODE REVIEW

File: sonic-db-cli/tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#include "sonic-db-cli.h"
+
+std::string db_file = "./database_config.json";
+std::string global_db_file = "./database_global.json";
+
+#define TEST_DB  "APPL_DB"
+#define TEST_NAMESPACE  "asic0"
+#define INVALID_NAMESPACE  "invalid"
+
+class SwsscommonEnvironment : public ::testing::Environment {
```
Comment (liuh-80) [inline sonic-db-cli/tests/main.cpp:15]: Still need add more E2E test to validate this new cli can run redis command correctly.

File: sonic-db-cli/Makefile.am [hunk 2]
```diff
@@ -0,0 +1,14 @@
+INCLUDES = -I $(top_srcdir)
+
+if DEBUG
+DBGFLAGS = -ggdb -DDEBUG
+else
+DBGFLAGS = -g -DNDEBUG
+endif
+
+bin_PROGRAMS = sonic-db-cli
+sonic_db_cli_SOURCES = sonic-db-cli.h sonic-db-cli.cpp
```
Comment (qiluo-msft) [inline sonic-db-cli/Makefile.am:10]: >sonic-db-cli.h [](http://example.com/codeflow?start=23&length=14)

Curios whether need to include *.h here?

File: sonic-db-cli/sonic-db-cli.cpp [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    po::store(po::command_line_parser(argc, argv)
+                 .options(allOptions)
+                 .positional(positional_opts)
+                 .run(),
+             variablesMap);
+    po::notify(variablesMap);
+}
+
+#ifdef TESTING
+int sonic_db_cli(int argc, char** argv)
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:231]: >sonic_db_cli [](http://example.com/codeflow?start=4&length=12)

You may define it as standalone function and main just call it. It will simplify `#ifdef` in header.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 4]
```diff
@@ -0,0 +1,278 @@
+#include <iostream>
+#include "sonic-db-cli.h"
+
+using namespace swss;
+using namespace std;
+
+static string g_netns;
+static string g_dbOrOperation;
+const char* emptyStr = "";
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:9]: >emptyStr [](http://example.com/codeflow?start=12&length=8)

Use EMPTY_STRING for a const name. However, I think use `""` in code without a named const may be more readable.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    cout << "Example 1: sonic-db-cli -n asic0 CONFIG_DB keys \\*" << endl;
+    cout << "Example 2: sonic-db-cli -n asic2 APPL_DB HGETALL VLAN_TABLE:Vlan10" << endl;
+    cout << "Example 3: sonic-db-cli APPL_DB HGET VLAN_TABLE:Vlan10 mtu" << endl;
+    cout << "Example 4: sonic-db-cli -n asic3 APPL_DB EVAL \"return {KEYS[1],KEYS[2],ARGV[1],ARGV[2]}\" 2 k1 k2 v1 v2" << endl;
+    cout << "Example 5: sonic-db-cli PING | sonic-db-cli -s PING" << endl;
+    cout << "Example 6: sonic-db-cli SAVE | sonic-db-cli -s SAVE" << endl;
+    cout << "Example 7: sonic-db-cli FLUSHALL | sonic-db-cli -s FLUSHALL" << endl;
+}
+
+void printRedisReply(redisReply* reply)
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:27]: >printRedisReply [](http://example.com/codeflow?start=5&length=15)

I do not find corresponding function in original python sonic-db-cli. If possible, translate function by function. Refactor only when there is a strong reason.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        throw runtime_error("Unexpected reply type");
+    }
+
+    cout << endl;
+}
+
+string buildRedisOperation(vector<string>& commands)
+{
+    int argc = (int)commands.size();
+    const char** argv = new const char*[argc];
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:57]: >argv [](http://example.com/codeflow?start=17&length=4)

Try use `vector<string>` to automate memory management. Try prevent new/delete/free/strdup usage. Extreme performance is non-goal in this PR.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        free(const_cast<char*>(argv[i]));
+    }
+    delete[] argv;
+    delete[] argvc;
+
+    return string(command.c_str());
+}
+
+int connectDbInterface(
+    DBInterface& dbintf,
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:79]: >DBInterface [](http://example.com/codeflow?start=4&length=11)

C++ DBInterface was introduced to mimic python implementation. Better not to use it in pure C++ application. Suggest using DBConnector.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+            auto db_socket = SonicDBConfig::getDbSock(db_name);
+            dbintf.set_redis_kwargs(db_socket, "", 0);
+        }
+
+        int db_id =  SonicDBConfig::getDbId(db_name, netns);
+        dbintf.connect(db_id, db_name, true);
+    }
+    catch (const exception& e)
+    {
+        cerr << "Invalid database name input: " << db_name << endl;
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:103]: >: [](http://example.com/codeflow?start=44&length=1)

Regarding the blank style, even the original is considered not perfect, let's do not change behavior in this PR.

We may need to unit test both to make sure they behave the same way.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+int handleAllInstances(
+    const string& netns,
+    const string& operation,
+    bool isTcpConn)
+{
+    auto db_names = SonicDBConfig::getDbList(netns);
+    for (auto& db_name : db_names)
+    {
+        int result = handleSingleOperation(netns, db_name, operation, isTcpConn);
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:150]: >handleSingleOperation [](http://example.com/codeflow?start=21&length=21)

Original implementation has `Pool` optimization, which is significant for `SAVE` case, which is critical during warm shutdown path.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+            // Use the tcp connectivity if namespace is local and unixsocket cmd_option is present.
+            isTcpConn = true;
+        }
+        
+        if (variablesMap.count("cmd"))
+        {
+            auto commands = variablesMap["cmd"].as< vector<string> >();
+            return executeCommands(dbOrOperation, commands, netns, isTcpConn);
+        }
+        else if (netns.compare("PING") == 0 || netns.compare("SAVE") == 0 || netns.compare("FLUSHALL") == 0)
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:185]: >compare [](http://example.com/codeflow?start=23&length=7)

`==` may be more readable.

File: debian/sonic-db-cli.install [hunk 11]
```diff
@@ -0,0 +1 @@
+usr/bin/sonic-db-cli
```
Comment (qiluo-msft) [inline debian/sonic-db-cli.install:1]: >sonic-db-cli [](http://example.com/codeflow?start=8&length=12)

Customer said they tried this solution, but still python sonic-db-cli is invoked.

Could you check if `hash -r` will fix the issue?
Or do we need another PR to remove python version completely?

File: sonic-db-cli/sonic-db-cli.cpp (hunk 5 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp:27]: When python code run redis command, it call print method to print the output:
https://github.com/Azure/sonic-py-swsssdk/blob/2cd6236849d194625c014d1c2c370f6d2513eb76/src/swsssdk/scripts/sonic-db-cli#:~:text=print(%27%5Cn%27.join(msg))

however, because c++ can't print redis command result directly, so we need add this new method.

File: sonic-db-cli/Makefile.am (hunk 2 above)
Comment (liuh-80) [reply sonic-db-cli/Makefile.am]: Fixed, removed this .h file here.

File: sonic-db-cli/sonic-db-cli.cpp (hunk 3 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp:253]: Fixed, add a main.cpp for main(), also because we need link most code to UT for test, so also add a new lib: libsoniccli.so

File: sonic-db-cli/sonic-db-cli.cpp (hunk 4 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp]: Fixed.

File: sonic-db-cli/sonic-db-cli.cpp (hunk 6 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp]: Fixed.

File: sonic-db-cli/sonic-db-cli.cpp (hunk 7 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp]: Fixed, now using DBConnector

File: sonic-db-cli/sonic-db-cli.cpp (hunk 8 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp:91]: Will discussion offline

File: sonic-db-cli/sonic-db-cli.cpp (hunk 9 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp:134]: Not using parallel std::for_each because it's support by c++17, will dicussion offline about if build swsscommon with c++17 or use other solution:

#include <execution>

    std::for_each(
        std::execution::par,
        db_names.begin(),
        db_names.end(),
        [=](auto&& db_name)
        {
            handleSingleOperation(netns, db_name, operation, isTcpConn);
        });

File: sonic-db-cli/sonic-db-cli.cpp (hunk 10 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp:163]: Fixed

File: debian/sonic-db-cli.install (hunk 11 above)
Comment (liuh-80) [reply debian/sonic-db-cli.install:1]: Yes, we need a PR to remove python version, the location is /usr/local/bin/sonic-db-cli

File: sonic-db-cli/sonic-db-cli.cpp (hunk 5 above)
Comment (liuh-80) [reply sonic-db-cli/sonic-db-cli.cpp:29]: Move this method to redis reply class as tostring()

File: sonic-db-cli/sonic-db-cli.cpp (hunk 10 above)
Comment (qiluo-msft) [reply sonic-db-cli/sonic-db-cli.cpp:163]: I do not see any changes. I am proposing
```
netns == "PING"
```

File: tests/Makefile.am [hunk 12]
```diff
@@ -36,8 +36,9 @@ tests_SOURCES = redis_ut.cpp                \
                 status_code_util_test.cpp   \
                 saiaclschema_ut.cpp         \
                 timer_ut.cpp                \
+                cli_ut.cpp                  \
                 main.cpp
 
 tests_CFLAGS = $(DBGFLAGS) $(AM_CFLAGS) $(CFLAGS_COMMON) $(CFLAGS_GTEST) $(LIBNL_CFLAGS)
 tests_CPPFLAGS = $(DBGFLAGS) $(AM_CFLAGS) $(CFLAGS_COMMON) $(CFLAGS_GTEST) $(LIBNL_CFLAGS)
-tests_LDADD = $(LDADD_GTEST) -lpthread -L$(top_srcdir)/common -lswsscommon $(LIBNL_LIBS) $(CODE_COVERAGE_LIBS)
+tests_LDADD = $(LDADD_GTEST) -lpthread -L$(top_srcdir)/common -lswsscommon $(LIBNL_LIBS) $(CODE_COVERAGE_LIBS)  -L$(top_srcdir)/sonic-db-cli -lsonicdbcli
```
Comment (qiluo-msft) [inline tests/Makefile.am:44]: >   [](http://example.com/codeflow?start=110&length=2)

One blank

File: debian/sonic-db-cli.install (hunk 11 above)
Comment (dprital) [reply debian/sonic-db-cli.install:1]: @liuh-80 , can you please contact me (drorp@nvidia.com) for offline discussion about this PR ?

**Comment by dprital** [comment]
@liuh-80  - As we would like to take this PR to 202111, there is a conflict need to be addressed:

tests/Makefile.am
diff --cc tests/Makefile.am
index a84a6df,1f0001f..0000000
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@@ -35,6 -35,8 +35,11 @@@ tests_SOURCES = redis_ut.cp
                  boolean_ut.cpp              \
                  status_code_util_test.cpp   \
                  saiaclschema_ut.cpp         \
++<<<<<<< ours
++=======
+                 timer_ut.cpp                \
+                 cli_ut.cpp                  \
++>>>>>>> theirs
                  main.cpp
  

" timer_ut.cpp" exist on master but not on 202111.

**Comment by liuh-80** [comment]
> @liuh-80 - As we would like to take this PR to 202111, there is a conflict need to be addressed:
> 
> tests/Makefile.am diff --cc tests/Makefile.am index a84a6df,1f0001f..0000000 --- a/tests/Makefile.am +++ b/tests/Makefile.am @@@ -35,6 -35,8 +35,11 @@@ tests_SOURCES = redis_ut.cp boolean_ut.cpp status_code_util_test.cpp saiaclschema_ut.cpp ++<<<<<<< ours ++=======
> 
> * ```
>               timer_ut.cpp                \
>   ```
> * ```
>               cli_ut.cpp                  \
>   ```
> 
> ++>>>>>>> theirs main.cpp
> 
> " timer_ut.cpp" exist on master but not on 202111.

This PR will submit to master branch first, then cherry pick to 202111 branch, the conflict will fix during cherry pick.

**Comment by qiluo-msft** [comment]
@liuh-80 could you provide an ETA of this PR?

**Comment by liuh-80** [comment]
> @liuh-80 could you provide an ETA of this PR passing functionality test?

ETA will be end of this month, I'm doing functionality test with this PR, which will remove old CLI and make sure new CLI pass all E2E test cases:
https://github.com/Azure/sonic-buildimage/pull/10825

File: common/redisreply.h [hunk 13]
```diff
@@ -92,9 +93,12 @@ class RedisReply
     /* Check that the status is QUEUED, throw exception otherwise */
     void checkStatusQueued();
 
+    std::string toString();
```
Comment (qiluo-msft) [inline common/redisreply.h:96]: >toString [](http://example.com/codeflow?start=16&length=8)

I see other files prefer `to_string()`, trying to mimic `std:to_string(...)`. Could you follow?

File: common/redisreply.h [hunk 14]
```diff
@@ -2,6 +2,7 @@
 #define __REDISREPLY__
 
 #include <hiredis/hiredis.h>
+#include <sstream>
```
Comment (qiluo-msft) [inline common/redisreply.h:5]: >sstream [](http://example.com/codeflow?start=10&length=7)

Not used in header file, you may move it to cpp file.

File: debian/sonic-db-cli.dirs [hunk 15]
```diff
@@ -0,0 +1 @@
+usr/local/bin
```
Comment (qiluo-msft) [inline debian/sonic-db-cli.dirs:1]: What is your migration plan? We discuss two strategy:
1. install to `/usr/bin`, and raise another small PR on buildimage to switch which package to install
2. install to `/usr/loca/bin`, and when move submodule, it will overwrite python version on buildimage.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+    catch (invalid_argument const& e)
+    {
+        cerr << "Command Line Error: " << e.what() << endl;
+        printUsage();
+        return -1;
+    }
+    catch (logic_error const& e)
+    {
+        // getopt_long throw logic_error when found a unknown option without value.
+        cerr << "Unknown option without value: " << endl;
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:255]: >: [](http://example.com/codeflow?start=45&length=1)

Do you want to append `e.what()`?

File: sonic-db-cli/sonic-db-cli.cpp [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+        auto netns = options.m_namespace;
+        bool isTcpConn = !options.m_unixsocket;
+        // Load the database config for the namespace
+        if (netns != "None" && !netns.empty())
+        {
+            // SonicDBConfig may initialized when run cli with UT
+            if (!SonicDBConfig::isGlobalInit())
+            {
+                SonicDBConfig::initializeGlobalConfig(global_config_file);
+            }
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:278]: Something is wrong in unit test. We should not workaround in the feature code.

File: sonic-db-cli/sonic-db-cli.cpp [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+        return 0;
+    }
+
+    if (!options.m_db_or_op.empty())
+    {
+        auto dbOrOperation = options.m_db_or_op;
+        auto netns = options.m_namespace;
+        bool isTcpConn = !options.m_unixsocket;
+        // Load the database config for the namespace
+        if (netns != "None" && !netns.empty())
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.cpp:272]: >None [](http://example.com/codeflow?start=22&length=4)

Why check with `None` string?

File: sonic-db-cli/sonic-db-cli.h [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    const std::string& operation,
+    bool isTcpConn);
+
+void parseCliArguments(
+    int argc,
+    char** argv,
+    Options &options);
+
+int sonic_db_cli(
+    const std::string &config_file,
```
Comment (qiluo-msft) [inline sonic-db-cli/sonic-db-cli.h:51]: >config_file [](http://example.com/codeflow?start=23&length=11)

Add default value `DEFAULT_SONIC_DB_CONFIG_FILE` ?

File: sonic-db-cli/sonic-db-cli.cpp (hunk 18 above)
Comment (qiluo-msft) [reply sonic-db-cli/sonic-db-cli.cpp:280]: This comment is still applicable.

File: common/redisreply.cpp [hunk 20]
```diff
@@ -158,7 +159,7 @@ void RedisReply::checkReplyType(int expectedType)
         const char *err = (m_reply->type == REDIS_REPLY_STRING || m_reply->type == REDIS_REPLY_ERROR) ?
             m_reply->str : "NON-STRING-REPLY";
 
-        string errmsg = "Expected to get redis type " + to_string(expectedType) + " got type " + to_string(m_reply->type) + ", err: " + err;
+        string errmsg = "Expected to get redis type " + std::to_string(expectedType) + " got type " + std::to_string(m_reply->type) + ", err: " + err;
```
Comment (qiluo-msft) [inline common/redisreply.cpp:162]: >std:: [](http://example.com/codeflow?start=56&length=5)

I am not sure if it will introduce ambiguity. Seems like you can remove `std::`.
Not a big issue.

**Comment by liuh-80** [comment]
After integrate this change to sonic-buildimage repo, all E2E test passed: https://github.com/Azure/sonic-buildimage/pull/10825/checks

**Review by qiluo-msft — APPROVED** [review]
